### PR TITLE
Return partial message when unpack errors

### PR DESCRIPTION
**Pull request by meparle**
https://github.com/moov-io/iso8583/issues/296

### CODE REVIEW

File: message.go [hunk 1]
```diff
@@ -249,7 +250,7 @@ func (m *Message) wrappErrorUnpack(src []byte) error {
 // unpack contains the core logic for unpacking the message. This method does
 // not handle locking or error wrapping and should typically be used internally
 // after ensuring concurrency safety.
-func (m *Message) unpack(src []byte) error {
+func (m *Message) unpack(src []byte) (string, error) {
```
Comment (cheukwing) [inline message.go:253]: Could we just create the `UnpackError` in `unpack`, and remove `wrappErrorUnpack`?
It doesn't look like `unpack` is used anywhere else in this file, and it would make the method signature clearer

File: message_test.go [hunk 2]
```diff
@@ -751,6 +751,61 @@ func TestPackUnpack(t *testing.T) {
...
... 0x30, 0x32, 0x6, 0x43, 0x39, 0x30, 0x31, 0x30, 0x32, 0x30, 0x36, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x33, 0x37, 0x34, 0x32, 0x37, 0x36, 0x35, 0x35, 0x35, 0x35, 0x35, 0x35, 0x35, 0x35, 0x35, 0x35, 0x35, 0x35, 0x3d, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30, 0x39, 0x38, 0x37, 0x36, 0x35, 0x34, 0x33, 0x32, 0x31, 0x30, 0x30, 0x31, 0x30, 0x30, 0x30, 0x30, 0x30, 0x33, 0x32, 0x31, 0x31, 0x32, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x33, 0x34, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x54, 0x65, 0x73, 0x74, 0x20, 0x74, 0x65, 0x78, 0x74, 0x64, 0x30, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7, 0x8, 0x31, 0x32, 0x33, 0x34, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x32, 0x33, 0x9a, 0x6, 0x32, 0x31, 0x30, 0x37, 0x32, 0x30, 0x9f, 0x2, 0xc, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x35, 0x30, 0x31, 0x30, 0x31, 0x37, 0x41, 0x6e, 0x6f, 0x74, 0x68, 0x65, 0x72, 0x20, 0x74, 0x65, 0x73, 0x74, 0x20, 0x74, 0x65, 0x78}
```
Comment (cheukwing) [inline message_test.go:758]: worth adding a small comment about why this should fail

File: message.go [hunk 3]
```diff
@@ -500,13 +503,11 @@ func (m *Message) Unmarshal(v interface{}) error {
 			if dataField.IsNil() && dataField.Kind() != reflect.Slice {
 				dataField.Set(reflect.New(dataField.Type().Elem()))
 			}
-			err := messageField.Unmarshal(dataField.Interface())
-			if err != nil {
+			if err := messageField.Unmarshal(dataField.Interface()); err != nil {
```
Comment (cheukwing) [inline message.go]: meganit: generally I would avoid driveby changes in an open source repo so that the change history is clearer as each PR has a single focus

**Review by cheukwing — APPROVED** [review]
lgtm, happy for you to setup a PR to the moov repo

File: errors.go [hunk 4]
```diff
@@ -4,6 +4,7 @@ package iso8583
 // connection failed to unpack message
 type UnpackError struct {
 	Err        error
+	Field      string
 	RawMessage []byte
 }
```
Comment (kelvinjhwong) [inline errors.go:10]: Is it worth updating `Error()` method to also interpolate the `Field`? But don't know whether this might be seen as a breaking change

**Review by kelvinjhwong — APPROVED** [review]
Looks good! Much simpler than the previous attempt indeed :raised_hands:

**Comment by alovak** [comment]
Hey! I think this PR looks good!

How do you plan to use the `Field` field from the `UnpackError`? 

Also, maybe we can use `FieldIndex` or `FieldID` or `FailedFieldID` as the name of the field?

**Comment by meparle** [comment]
> How do you plan to use the `Field` field from the `UnpackError`?

This will go into DE44 where we indicate the field number which failed parsing, in our response to the scheme when we return with a Format Error response code. Having the field ID separately means not needing to extract it from the error message.
 
> Also, maybe we can use `FieldIndex` or `FieldID` or `FailedFieldID` as the name of the field?

Sure thing - I'll change it to `FieldID`.

**Comment by cheukwing** [comment]
this has been merged to upstream master
